portfolio-project-matching-app/Firebase/auth.js

```diff
@@ -1,36 +1,133 @@
-import { getAuth, signInWithPopup, GoogleAuthProvider, onAuthStateChanged, signOut, auth } from "firebase/auth";
+import { getAuth, 
+    signInWithPopup, 
+    GoogleAuthProvider, 
+    createUserWithEmailAndPassword,
+    signInWithEmailAndPassword,
+} from "firebase/auth";
 import { createNewUserDocWithId, getUserById } from '../backend/daoUser';
-import { useAuthUpdate } from '../context/AuthContext';
-import { firebaseApp } from "./clientApp.ts";
 
 
 
 const signinWithGoogle = async() => {
+    /* Signs the user in with google.
+        
+    User must already have an account.
+        
+    If the user doesn't have an account or firebase throws an error, 
+    this function will throw an error.
+    */
     const provider = new GoogleAuthProvider();
     const auth = getAuth();
     try {
         const result = await signInWithPopup(auth, provider);
         const user = result.user;
         // Check if user has a doc collection in the database
+        console.log(`${JSON.stringify(user)}`)
+        let userDoc = await getUserById(user.uid);
+        if (userDoc === -1){
+            // If the user is here for the first time, throw an error
+            throw({code: "auth/no-user-found"})
+        }
+        return userDoc;
+    } catch (error) {
+        throw(error);
+    }    
+}
+
+const signupWithGoogle = async() => {
+    /* Signs the user up with google.
+        
+    User must not yet have an account.
+        
+    If the user does have an account or firebase throws an error, 
+    this function will throw an error.
+    */
+    const provider = new GoogleAuthProvider();
+    const auth = getAuth();
+    try {
+        const result = await signInWithPopup(auth, provider);
+        const user = result.user;
+        // Check if user has a doc collection in the database
+        console.log(`${JSON.stringify(user)}`)
         let userDoc = await getUserById(user.uid);
         if (userDoc === -1){
             // If the user is here for the first time, make them a new user doc
-            const newUser = await createNewUserDocWithId({
+            await createNewUserDocWithId({
                 // Some default values for their username, email, and introduction
                 email: user.email,
                 username: user.email,
                 introduction: `Hi, I'm ${user.displayName}`,
             }, user.uid);
             userDoc = await getUserById(user.uid);
+        } else {
+            throw({code: 'auth/email-already-in-use'})
+        }
+        return userDoc;
+    } catch (error) {
+        throw(error);
+    }    
+}
+
+const signUpWithEmailAndPassword = async(email, password) => {
+    /* Signs the user in with username and password.
+        
+    User must not have an account.
+        
+    If the user doesn have an account or firebase throws an error, 
+    this function will throw an error.
+    */
```

Comment:
This is awesome. I think it is great that our project has this capability.

Reply:
Yeah, I'm glad it does too. I was thinking the implementation would be more challenging. Thanks Firebase!